ouroboros-consensus/src/Ouroboros/Storage/ImmutableDB/Impl/Validation.hs

```diff
@@ -373,6 +358,38 @@ validateEpoch ValidateEnv{..} shouldBeFinalised epoch mbPrevHash = do
         UnexpectedError (e@InvalidFileError {}) -> Just e
         _ -> Nothing) err
 
+    -- | When reading the entries from the secondary index file, we need to
+    -- pass in a value of type 'IsEBB' so we know whether the first entry
+    -- corresponds to an EBB or a regular block. We need this information to
+    -- correctly interpret the deserialised 'Word64' as a 'BlockOrEBB': if
+    -- it's an EBB, it is the 'EpochNo' ('Word64'), it's a regular block, is a
+    -- 'SlotNo' ('Word64').
+    --
+    -- However, at the point we are reading the secondary index file, we don't
+    -- yet know whether the first block will be an EBB or a regular block. We
+    -- will find that out when we read the actual block from the epoch file.
+    --
+    -- Fortunately, we can make a /very/ good guess: if the 'Word64' of the
+    -- 'BlockOrEBB' matches the epoch number, it is almost certainly an EBB,
+    -- as the slot numbers increase @10k@ times faster than epoch numbers.
+    -- E.g., the first slot number in epoch 1 will be 21600 if @k = 2160@. We
+    -- could only make the wrong guess in the first very first epoch, i.e.,
+    -- epoch 0, as the first slot number is also 0. However, we know that the
+    -- real blockchain starts with an EBB, so even in that case we're fine.
```

Comment:
Property: for every epoch `e`, for all slot numbers `s` in the epoch we have `s > e`.

Reply:
We need a stronger consequence (follows from this property): it's an EBB _if and only if_ it matches (because you are actually depending on the opposite direction in `fixupEBB`.